packages/tailwindcss/src/canonicalize-candidates.ts

```diff
@@ -0,0 +1,1510 @@
+          )) {
+            yield replacementCandidate
+          }
+        }
+      }
+    }
+  }
+}
+
+// Let's make sure that all variables used in the value are also all used in the
+// found replacement. If not, then we are dealing with a different namespace or
+// we could lose functionality in case the variable was changed higher up in the
+// DOM tree.
+function allVariablesAreUsed(
+  designSystem: DesignSystem,
+  candidate: Candidate,
+  replacement: Candidate,
+) {
+  let value: string | null = null
+
+  // Functional utility with arbitrary value and variables
+  if (
+    candidate.kind === 'functional' &&
+    candidate.value?.kind === 'arbitrary' &&
+    candidate.value.value.includes('var(--')
+  ) {
+    value = candidate.value.value
+  }
+
+  // Arbitrary property with variables
+  else if (candidate.kind === 'arbitrary' && candidate.value.includes('var(--')) {
+    value = candidate.value
+  }
+
+  // No variables in the value, so this is a safe migration
+  if (value === null) {
+    return true
+  }
+
+  let replacementAsCss = designSystem
+    .candidatesToCss([designSystem.printCandidate(replacement)])
+    .join('\n')
+
+  let isSafeMigration = true
+  ValueParser.walk(ValueParser.parse(value), (node) => {
+    if (node.kind === 'function' && node.value === 'var') {
+      let variable = node.nodes[0].value
+      let r = new RegExp(`var\\(${variable}[,)]\\s*`, 'g')
+      if (
+        // We need to check if the variable is used in the replacement
+        !r.test(replacementAsCss) ||
+        // The value cannot be set to a different value in the
+        // replacement because that would make it an unsafe migration
+        replacementAsCss.includes(`${variable}:`)
+      ) {
+        isSafeMigration = false
+        return ValueParser.ValueWalkAction.Stop
+      }
+    }
+  })
+
+  return isSafeMigration
+}
+
+// ----
+
+function bareValueUtilities(designSystem: DesignSystem, candidate: Candidate): Candidate {
+  // We are only interested in bare value utilities
+  if (candidate.kind !== 'functional' || candidate.value?.kind !== 'named') {
+    return candidate
+  }
+
+  let utilities = preComputedUtilities.get(designSystem)
+  let signatures = computeUtilitySignature.get(designSystem)
+
+  let targetCandidateString = designSystem.printCandidate(candidate)
+
+  // Compute the signature for the target candidate
+  let targetSignature = signatures.get(targetCandidateString)
+  if (typeof targetSignature !== 'string') return candidate
+
+  // Try a few options to find a suitable replacement utility
+  for (let replacementCandidate of tryReplacements(targetSignature, candidate)) {
+    let replacementString = designSystem.printCandidate(replacementCandidate)
+    let replacementSignature = signatures.get(replacementString)
+    if (replacementSignature !== targetSignature) {
+      continue
+    }
+
+    // Update the candidate with the new value
+    replaceObject(candidate, replacementCandidate)
```

Comment:
I think all the uses of `replaceObject(candidate, replacementCandidate); return replacementCandidate` can be just `return replacementCandidate` right?

It'd be cool if we could get rid of `replaceObject` entirely but that's probably not realistic.